src/core/lib/security/credentials/tls/grpc_tls_certificate_provider.cc

```diff
@@ -413,6 +486,17 @@ absl::StatusOr<bool> PrivateKeyAndCertificateMatch(
   return result;
 }
 
+absl::Status IsKeyCertPairsListValid(const PemKeyCertPairList& pair_list) {
+  for (size_t i = 0; i < pair_list.size(); ++i) {
+    absl::StatusOr<bool> matched_or = PrivateKeyAndCertificateMatch(
+        pair_list[i].private_key(), pair_list[i].cert_chain());
+    if (!(matched_or.ok() && *matched_or)) {
+      return matched_or.status();
```

Comment:
Sorry I missed this. But what happens if the status was ok and matching failed?
Here we are returning OK status, which is incorrect. You would need to create a status here for failed match.

Reply:
That's a great observation. Noted.

Reply:
I feel we should propagate the errors from PrivateKeyAndCertificateMatch function. 
You could do
```
if(!matched_or.ok()){
  return matched_or.status();
}
if(!*matched_or) {
  return absl::InvalidArgumentError("...");
}
```

Reply:
You could take a look at https://github.com/grpc/grpc/pull/26663/commits/3794c98612b80c0ade7de9dc66514de31824e60f where I have changed the logic.

Reply:
That is unnecessary nesting. Please check out the suggested snippet above.

Reply:
It's not unnecessary. From the [guide](https://abseil.io/docs/cpp/guides/status),
"accessing the object held by an absl::StatusOr<T> should be performed via operator* or operator->, after a call to ok() confirms that the absl::StatusOr<T> holds an object of type T"

Reply:
By nesting, I was referring to if-if nesting. That is unnecessary and complicates a very simple logic. Please update the code.